torch/testing/_internal/opinfo/definitions/nested.py

```diff
@@ -0,0 +1,277 @@
+        yield nt
+
+        # without min / max seqlen cached
+        values = nt.values().clone().detach()
+        offsets = nt.offsets().clone().detach()
+        yield torch.nested.nested_tensor_from_jagged(values, offsets)
+
+    # TODO: add non-contiguous NJTs
+
+
+# Computes an unbind-based reference for a given OpInfo on a given SampleInput.
+# This reference unbinds the input NJT and invokes the op on each of the components,
+# optionally wrapping the result in an NJT.
+def unbind_reference(op, sample, wrap_output_as_njt=True):
+    assert sample.input.is_nested
+    out_ref_components = []
+    for i, component in enumerate(sample.input.unbind(dim=0)):
+
+        def _slice_njts(t, i=i, inp=sample.input):
+            # any NJT with the same ragged structure as the input should
+            # also be sliced to pass to the reference
+            if isinstance(t, torch.Tensor) and _raggedness_matches(t, inp):
+                return t[i]
+            else:
+                return t
+
+        args = tree_map(_slice_njts, sample.args)
+        kwargs = tree_map(_slice_njts, sample.kwargs)
+
+        from torch._prims_common import canonicalize_dims
+
+        # Need to adjust dim to apply on NJT component
+        if "dim" in kwargs:
+            kwargs["dim"] = canonicalize_dims(sample.input.dim(), kwargs["dim"]) - 1
+            assert kwargs["dim"] >= 0
+
+        # TODO: handle this
+        assert "dims" not in kwargs
+
+        out_ref_component = op.op(component, *args, **kwargs)
+
+        # TODO: handle list / tuple / non-NJT outputs
+        assert not isinstance(out_ref_component, (list, tuple))
+        out_ref_components.append(out_ref_component)
+
+    if wrap_output_as_njt:
+        return torch.nested.as_nested_tensor(out_ref_components, layout=torch.jagged)
+
+    return out_ref_components
+
+
+# Computes the reference value for a reduction op.
+def reduction_reference(op, sample):
+    assert sample.input.is_nested
+    dim = sample.kwargs.get("dim", None)
+    keepdim = sample.kwargs.get("keepdim", False)
+    assert dim != 0, "reductions over the batch dim are not supported"
+    assert "dims" not in sample.kwargs
+    assert sample.input._ragged_idx == 1
+
+    if dim is None:
+        # calculate reference value by running reduction on values buffer
+        return op.op(sample.input.values(), *sample.args, **sample.kwargs)
+
+    if dim == sample.input._ragged_idx:
+        # calculate reference value by running an unbind reference and stacking
+        out_ref_components = unbind_reference(op, sample, wrap_output_as_njt=False)
+        return torch.stack(out_ref_components, dim=0)
+
+    # unbind reference works for other reductions
+    return unbind_reference(op, sample)
+
+
+def sample_inputs_njt_general(op_info, device, dtype, requires_grad, **kwargs):
+    for njt in _sample_njts(
+        device=device, dtype=dtype, requires_grad=requires_grad, dims=[2, 3, 4]
+    ):
+        yield SampleInput(njt)
```

Comment:
I can imagine this being supplemented with one or two test cases that test that non-tensor args are properly propagated